### ISSUE TITLE
How to interpret model output, and get X year disease risk?

### ISSUE THREAD

**Issue by RobertDaviesGenomics** [gerstung-lab/Delphi#5]
So I've got a bit of a handle on the code base but I want to make sure I'm interpreting the output correctly. 

So e.g. this code in 
[evaluate_delphi.ipynb](https://github.com/gerstung-lab/Delphi/blob/main/evaluate_delphi.ipynb
)

I see code like this, which I can run
```
d100k = get_batch(range(dataset_subset_size), val, val_p2i,  
              select='left', block_size=128, 
              device=device, padding='random')
p100k = []
model.to(device)
batch_size=256
with torch.no_grad():
    for dd in tqdm(zip(*map(lambda x: torch.split(x, batch_size), d100k)), total=d100k[0].shape[0]//batch_size+1):
        p100k.append(model(*[x.to(device) for x in dd])[0].cpu().detach()[:,:,diseases_of_interest].numpy())
p100k = np.vstack(p100k)
```
and this gives `d100k` which contains reformatted time and event data for some samples, and `p100k` contains model output. 

Now I'm assuming this `p100k` contains some version of the logits described in the paper?

In the paper it says 
_Modelled incidence. In the exponential waiting time definition above, the logits of the model
correspond to the log probabilities of daily occurrences of each token. The probability of an
event occurring within a year is given by P(T < 365.25) = 1 - exp(exp(lambda) * 365.25)._

So my first question, does `p100k` contain column wise these logit outputs, corresponding one-to-one with the entries of `d100k` at the age of the sample as specified in the corresponding column of `d100k[1]`? Importantly, the last column would be the final model output, at the age of the sample when they're being predicted?

Now, how should I interpret them, and transform them to X year disease risk? 

So if it's true that "the logits of the model correspond to the log probabilities of daily occurrences of each token", then assuming this means any day conditional on an event not occuring i.e the first day, and the output of the model (contents of `p100k`) are these logits, and for an exponentially distributed random variable T measuring a daily rate with parameter lambda, then if logit = log(P(T < 1)), then lambda = -log(1 - exp(logit)), and so say 10 year risk should be 
P(T < 3650) = 1 - exp(log(1 - exp(logit)) * 3650)

this would be something like this 
```
vec = 1 - np.exp(np.log(1 - np.exp(p100k[:, 127, 7])) * 3650)      ## here 7 is for heart attacks                                                                                                                                            
vec.sort()                                                                                                                                                                                                     
vec[:5]                                                                                                                                                                                                        
vec[-5:] 
```
giving 
```
>>> array([0.0002175 , 0.00043499, 0.00043499, 0.00043499, 0.00065249],                                                                                                                                    
      dtype=float32)                                                                                                                                                                                           
>>> array([0.68793714, 0.70542514, 0.746665  , 0.7535177 , 0.7960815 ],                                                                                                                                        
      dtype=float32) 
```
which seems to make sense

Does this all seem right?

**Comment by RobertDaviesGenomics** [comment]
One thing I was just thinking of, how does this deal with conditions whose rate increases with age? So for example the rate of heart attacks increases with age, all else being held equal. But the Delphi model output (I think) is "log probabilities of daily occurrences of each token" which presumably reflects the instantaneous probability at their last input age? Is this / can this be accounted for?
